### PR TITLE
[Refactor] Changed the way ACE and ACRE function overrides work.

### LINKED ISSUE #166: [Change] Refactor function overrides for external addons.
By Cre, from https://github.com/CombinedArmsGaming/CAFE3/pull/162

As discussed, I'd modify the ACRE override to use two distinct functions: your override, and the original function (compiled in the traditional way in `CfgFunctions`. but using a soft-dependency). The functions config would then become

```cpp
#if __has_include("\idi\acre\addons\main\stringtable.xml")
    class acre_ace_interact
    {
        class overrides
        {
            file = "components\radio\acre";
            class radioListChildrenActions{};
            class radioListChildrenActions_original
            {
                file = "idi\acre\addons\ace_interact\fnc_radioListChildrenActions.sqf";
            };
        }
    }
#endif
```

Then we can omit lines 5-12 of your override:

```sqf
if (isNil "f_var_originalFunction_addRadioActionsToClass") then
{
    // Why isn't this easier to do in SQF?  Why does it have to be so brittle?  thanks lads
    f_var_originalFunction_addRadioActionsToClass = compile preprocessFileLineNumbers "idi\acre\addons\ace_interact\fnc_radioListChildrenActions.sqf";
};


private _children = _this call f_var_originalFunction_addRadioActionsToClass;
```

and just do:

```sqf
private _children = _this call acre_ace_interact_fnc_radioListChildrenActions_original;
```

_Originally posted by @Cre8or in https://github.com/CombinedArmsGaming/CAFE3/issues/162#issuecomment-1591118556_

### PR DESCRIPTION
**Pull request by Bubbus**
Resolves #166.

### Pull Request Description
**When merged this pull request will:**
- Refactors ACE and ACRE overrides.
- ACRE override remains gated by ACRE presence.  ACE override not gated because ACE is a hard dependency.

### Release Notes
Changed the way we override ACE and ACRE functions in the framework.

## IMPORTANT

- [x] Testing has been completed as neccessary, depending on the nature & impact of the changes.
- [x] The Release Notes section below must be filled out appropriately to explain the changes made in this PR. This section will be used in Framework Changelogs.
- [ ] If the contribution affects [the wiki](https://github.com/CombinedArmsGaming/CAFE3/wiki), please include your changes in this pull request so the wiki is consistently updated.
- [x] [Contribution Guidelines](https://github.com/CombinedArmsGaming/CAFE3/blob/release/contributing.md) are read, understood and applied.
- [x] Title of this PR uses our standard template `[Descriptor] - Add|Fix|Improve|Change|Make|Remove {changes}`.

### CODE REVIEW

File: components/functions.hpp [hunk 1]
```diff
@@ -68,14 +71,20 @@ class ace_gunbag
...
+        {
+            file = "components\radio\acre";
+            class radioListChildrenActions{};
+            class radioListChildrenActions_original
+            {
+                file = "idi\acre\addons\ace_interact\fnc_radioListChildrenActions.sqf";
+            };
+        }
+    }
+#endif
```
Comment (Cre8or) [inline components/functions.hpp:87]: *Neat.*

**Review by Cre8or — COMMENTED** [review]
I like this change, it's cleaner and easier to maintain.

It just occured to me we could extend this preprocessor method to catch changes to addon function names. They happen very rarely, but have happened in the past - maybe something to consider within the scope of this PR?

**Comment by Bubbus** [comment]
> I like this change, it's cleaner and easier to maintain.
> 
> It just occured to me we could extend this preprocessor method to catch changes to addon function names. They happen very rarely, but have happened in the past - maybe something to consider within the scope of this PR?

I'm not sure what we'd do with this if it happens - I guess we'd want to fail safely, add fallback behaviour?  Sounds good because we should be taking special care when we're overriding things.  Also sounds like creep, so I think let's put it in another ticket.

**Comment by Cre8or** [comment]
> I'm not sure what we'd do with this if it happens - I guess we'd want to fail safely, add fallback behaviour? Sounds good because we should be taking special care when we're overriding things. Also sounds like creep, so I think let's put it in another ticket.

Agreed on both points. We already have a pretty strong dependency on specific ACRE and ACE functions being present, to the point where undesirable things happen when they're sudenly not there anymore. I think cementing this dependency with proper error checks (think assertions) somewhere, ideally in an init function, would be an appropriate step. Not only would we be able to inform the mission maker better via error messages (this could be done in a number of ways), but we'd be able to catch these issue early - at the very least before a scheduled event.

Definitely worth shoving into another ticket, though. We'll have to give it a proper think in order to decide how to best approach it. Let's put that on the backburner.

**Review by Cre8or — APPROVED** [review]
🚢

**Pull request merged**